build_tools/cmake/iree_check_test.cmake

```diff
@@ -0,0 +1,153 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+include(CMakeParseArguments)
+
+# iree_check_test()
+#
+# Creates a test using iree-check-module for the specified source file.
+#
+# Mirrors the bzl rule of the same name.
+#
+# Parameters:
+# NAME: Name of the target
+# SRC: mlir source file to be compiled to an IREE module.
+# TARGET_BACKEND: target backend to compile for.
+# DRIVER: driver to run the module with.
+# ARGS: additional args to pass to iree-check-module. The driver and input
+#       file are passed automatically.
+function(iree_check_test)
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;SRC;TARGET_BACKEND;DRIVER"
+    "ARGS"
+    ${ARGN}
+  )
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  set(_MODULE_TARGET_NAME "${_RULE_NAME}_module")
+
+  iree_bytecode_module(
+    NAME
+      "${_MODULE_TARGET_NAME}"
+    SRC
+      "${_RULE_SRC}"
+    FLAGS
+      "-iree-mlir-to-vm-bytecode-module"
+      "--iree-hal-target-backends=${_RULE_TARGET_BACKEND}"
+    TESTONLY
+  )
+
+  # TODO(bf/146898896): Do this in iree_bytecode_module and avoid having to
+  # reach into the internals.
+  set(_GENERATED_MODULE_NAME "${_MODULE_TARGET_NAME}.module")
+  add_custom_target(
+    "${_NAME}" ALL
+     DEPENDS
+       "${_GENERATED_MODULE_NAME}"
+  )
+
+  string(REPLACE "_" "/" _PACKAGE_PATH ${_PACKAGE_NAME})                                                                                                    
+  set(_NAME_PATH "${_PACKAGE_PATH}:${_RULE_NAME}") 
+
+  add_test(
+    NAME
+      "${_NAME_PATH}"
+    COMMAND
+      "${CMAKE_SOURCE_DIR}/build_tools/cmake/run_test.${IREE_HOST_SCRIPT_EXT}"
+      "$<TARGET_FILE:iree_modules_check_iree-check-module>"
+      "--input_file=${CMAKE_CURRENT_BINARY_DIR}/${_GENERATED_MODULE_NAME}"
+      "--driver=${_RULE_DRIVER}"
+      "${_RULE_ARGS}"
+  )
+
+  set_property(TEST "${_NAME_PATH}" PROPERTY REQUIRED_FILES "${_GENERATED_MODULE_NAME}")
+  set_property(TEST "${_NAME_PATH}" PROPERTY ENVIRONMENT "TEST_TMPDIR=${_NAME}_test_tmpdir")
+  set_property(TEST "${_NAME_PATH}" PROPERTY LABELS "${_PACKAGE_PATH}")
+endfunction()
+
+
+# iree_check_test_suite()
```

Comment:
Hm.. `iree_check_test_suite` is what is always called from the outside, but the file is module is named `iree_check_test` and also has `iree_check_test_suite` function implemented... Maybe the module should be renamed or splited?

Reply:
This is currently consistent with bazel and with iree_lit_test. The idea was that the suite was a convenience for creating a bunch of these tests but the core functionality was just having such a test. iree_check_test_suite should be very simple, but then we get into cmake limitations... Also iree_check_test on its own is intended to be a "public" function

I'm not really sure what the naming conventions we're following here are. We've got some places where it's one file per function and some where it's not. I like having these functions in the same file, since they're closely related though.

Reply:
If it's consistent I am fine! Missed to check bazel, sorry.

Reply:
No worries. Consistency is good, but we can also challenge the thing it's consistent with. We could change both the bazel and cmake one. But I think not this PR.